server/src/updaters/thread-updaters.js

```diff
@@ -427,30 +429,65 @@ async function updateThread(
     nextThreadType !== oldThreadType ||
     nextParentThreadID !== oldParentThreadID
   ) {
-    savePromises.recalculatePermissionsChangeset = recalculateAllPermissions(
+    intermediatePromises.recalculatePermissionsChangeset = recalculateAllPermissions(
       request.threadID,
       nextThreadType,
     );
   }
   const {
     addMembersChangeset,
     recalculatePermissionsChangeset,
-  } = await promiseAll(savePromises);
+  } = await promiseAll(intermediatePromises);
 
-  const changeset = [];
+  const membershipRows = [];
+  const relationshipRows = [];
   if (recalculatePermissionsChangeset && newMemberIDs) {
-    changeset.push(
-      ...recalculatePermissionsChangeset.filter(
+    const {
+      membershipRows: recalculateMembershipRows,
+      relationshipRows: recalculateRelationshipRows,
+    } = recalculatePermissionsChangeset;
+    membershipRows.push(
+      ...recalculateMembershipRows.filter(
         rowToSave => !newMemberIDs.includes(rowToSave.userID),
       ),
     );
+    const parentMemberIDs = recalculateMembershipRows
+      .filter(rowToSave => !newMemberIDs.includes(rowToSave.userID))
+      .map(rowToSave => rowToSave.userID);
+    const newUserIDs = newMemberIDs.filter(
+      memberID =>
+        !recalculateMembershipRows.find(
+          rowToSave => memberID === rowToSave.userID,
+        ),
+    );
```

Comment:
I think there is one more edge case that can break this, but the good news is that it should be easy to fix!

## Changing thread parent while simultaneously adding user that was in old parent

1. Thread A exists with users Beth and Joe
2. Thread B exists with users Bob and Joe
3. Thread C exists with single user Joe and is an open child of Thread A
4. Joe calls `updateThread` on Thread C with two changes: add Beth, and also move it to be an open child of Thread B
5. `recalculateAllPermissions` returns an `{ operation: 'delete', user: 'Beth', thread: 'C' }`, because it sees that Beth used to have permissions for Thread C, but doesn't anymore
6. We need to create a `know_of` between Beth and Bob. But Beth will get filtered out of `newUserIDs` because of the aforementioned `{ operation: 'delete' }`

## Fix

I think we can just update this code to check for `operation` as well. I'd also like to check `threadID` - it doesn't make a difference, but it makes it more clear what we're doing: figuring out what users are new to `request.threadID`.

```js
const newUserIDs = newMemberIDs.filter(
  memberID =>
    !recalculateMembershipRows.find(
      rowToSave => rowToSave.userID === memberID &&
        rowToSave.threadID === request.threadID &&
        rowToSave.operation !== 'delete',
    ),
);
```

Reply:
Do you think it applies also to `createThread`? If we want to reuse this logic in `getParentThreadMembershipRowsForNewUsers` it should 🤔